doctr/models/recognition/parseq/pytorch.py

```diff
@@ -0,0 +1,503 @@
+        logits = self.head(self.decode(ys, features, mask, target_query=pos_queries))
+
+        return logits  # (N, max_length, vocab_size + 1)
+
+    def decode_non_autoregressive(self, features: torch.Tensor) -> torch.Tensor:
+        """Decode the given features at once"""
+        pos_queries = self.pos_queries[:, : self.max_length + 1].expand(features.size(0), -1, -1)
+        ys = torch.full((features.shape[0], 1), self.vocab_size + 1, dtype=torch.long, device=features.device)
+        return self.head(self.decode(ys, features, target_query=pos_queries))[:, : self.max_length]
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        target: Optional[List[str]] = None,
+        return_model_output: bool = False,
+        return_preds: bool = False,
+    ) -> Dict[str, Any]:
+        features = self.feat_extractor(x)["features"]  # (batch_size, patches_seqlen, d_model)
+        # remove cls token
+        features = features[:, 1:, :]
+
+        if self.training and target is None:
+            raise ValueError("Need to provide labels during training")
+
+        if target is not None:
+            # Build target tensor
+            _gt, _seq_len = self.build_target(target)
+            gt, seq_len = torch.from_numpy(_gt).to(dtype=torch.long).to(x.device), torch.tensor(_seq_len).to(x.device)
+
+            if self.training:
+                # Generate permutations for the target sequences
+                tgt_perms = self.generate_permutations(seq_len)
+
+                # Create padding mask for target input
+                # [True, True, True, ..., False, False, False] -> False is masked
+                padding_mask = (
+                    ((gt != self.vocab_size + 2) | (gt != self.vocab_size)).unsqueeze(1).unsqueeze(1)
+                )  # (N, 1, 1, max_length)
+
+                for perm in tgt_perms:
+                    # Generate attention masks for the permutations
+                    _, target_mask = self.generate_permutations_attention_masks(perm)
+                    # combine target padding mask and query mask
+                    mask = (target_mask & padding_mask).int()
+                    logits = self.head(self.decode(gt, features, mask))  # (N, max_length, vocab_size + 1)
+            else:
+                # eval step - use non-autoregressive decoding while training evaluation
+                logits = self.decode_non_autoregressive(features)
+        else:
+            logits = self.decode_autoregressive(features)
+
+        out: Dict[str, Any] = {}
+        if self.exportable:
+            out["logits"] = logits
+            return out
+
+        if return_model_output:
+            out["out_map"] = logits
+
+        if target is None or return_preds:
+            # Post-process boxes
+            out["preds"] = self.postprocessor(logits)
+
+        if target is not None:
+            out["loss"] = self.compute_loss(logits, gt, seq_len, ignore_index=self.vocab_size + 2)
```

Comment:
This should be inside the loop where `logits` is computed. The model loss is the mean cross-entropy loss across all permutations. There's also a small detail about the computation of losses for `[E]` (where it is replaced by `[P]` after the forward and backward permutations). Please check the reference implementation:

https://github.com/baudm/parseq/blob/bc8d95cda4666d32fa53daf2ea97ff712b71e7c7/strhub/models/parseq/system.py#L242-L256

Reply:
Alright got it thanks 👍

Reply:
@nikokks this should be already handled by our ce_mask but i will check it next week